File: packages/core/src/model/part.ts
import type { PartMeasure } from "./measure";
import type { KitComponent } from "./kit";

/**
 * MNX interval — chromatic + diatonic components.
 */
export interface Interval {
  /** Chromatic distance in semitones (signed). */
  halfSteps: number;
  /** Diatonic staff distance in staff positions (signed). */
  staffDistance: number;
}

/**
 * MNX part-transposition — transposing instrument configuration.
 */
export interface Transposition {
  /** The sounding→written transposition interval. */
  interval: Interval;
  /** Circle-of-fifths value at which key signatures flip enharmonically. */
  keyFifthsFlipAt?: number;
  /** Instrument prefers written pitches even in concert-pitch scores. */
  prefersWrittenPitches?: boolean;
}

/**
 * Viritura vendor extensions for a Part (`_x.viritura`).
 *
 * These travel with the MNX score (under `_x.viritura` on the part) so
 * the editor and audio engine can resolve a Part to a stable instrument
 * identity without falling back to fuzzy name matching.
 */
export interface PartVirituraExt {
  /** Stable instrument-catalog ID (e.g. "flute", "bflat-clarinet"). */
  instrumentId?: string;
  /** General-MIDI program (0..127). Used directly by the audio engine. */
  midiProgram?: number;
  /** Instrument family for spatial placement / catalog routing. */
  family?: string;
  /** Spatial-audio stage position in concert-hall meters. Persisted from Play
   *  mode so a user's instrument arrangement survives reload. */
  spatial?: { x: number; y: number };
}

/**
 * A part (instrument) in the score (MNX "parts[n]").
 */
export interface Part {
  /** Unique part ID (referenced by layout sources) */
  id?: string;
  /** Part name (e.g., "Violin", "Flute 1") */
  name: string;
  /** Short name / abbreviation (e.g., "Vln.", "Fl. 1") */
  shortName?: string;
  /** Measures for this part (indexed same as global measures) */
  measures: PartMeasure[];
  /** Number of staves for this part (default 1; grand staff = 2, organ = 3) */
  staves?: number;
  /** Transposition for transposing instruments (MNX `transposition`). */
  transposition?: Transposition;
  /** Drum-kit component map (MNX `kit`). Key = component ID (referenced by
   *  kit-note `kitComponent`). When present, this is an unpitched percussion part. */
  kit?: Record<string, KitComponent>;
  /** Viritura vendor extensions (MNX `_x.viritura`). */
  _x?: { viritura?: PartVirituraExt };
}

File: engine/viritura-engine/src/model/part.rs
use super::kit::KitComponent;
use super::measure::PartMeasure;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// MNX interval — chromatic + diatonic components.
/// Convention: sounding pitch + interval = written pitch.
/// Ref: MNX spec objects/interval
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Interval {
    /// Chromatic distance in semitones (signed).
    pub half_steps: i32,
    /// Diatonic staff distance in staff positions (signed).
    pub staff_distance: i32,
}

/// MNX part-transposition — transposing instrument configuration.
/// Ref: MNX spec objects/part-transposition
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Transposition {
    /// The sounding→written transposition interval.
    pub interval: Interval,
    /// Circle-of-fifths value at which key signatures flip enharmonically.
    /// Non-negative: subtract 12 fifths; negative: add 12 fifths.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key_fifths_flip_at: Option<i32>,
    /// Instrument prefers written pitches even in concert-pitch scores
    /// (e.g. piccolo, glockenspiel, double bass).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prefers_written_pitches: Option<bool>,
}

/// A part (instrument) in the score.
///
/// Model-internal type — construction goes through `promote::part::promote_part`.
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct Part {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(default)]
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none", rename = "shortName")]
    pub short_name: Option<String>,
    pub measures: Vec<PartMeasure>,
    /// Number of staves for this part (default 1; grand staff = 2, organ = 3).
    pub staves: u32,
    /// Transposition for transposing instruments (MNX `transposition`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transposition: Option<Transposition>,
    /// Drum-kit component map (MNX `kit`). When present, this part is an
    /// unpitched percussion part. Keys are component IDs (e.g. "kick").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kit: Option<HashMap<String, KitComponent>>,
}
